rust/gitxetcore/src/git_integration/git_xet_repo.rs

```diff
@@ -1575,56 +1587,137 @@ impl GitXetRepo {
+                // Now, start the CAS (needed for all the steps below.
+                let cas = self.get_staging_cas().await?;
+
+                // Begin uploading all the CAS blocks to the remote.
+                let upload_all_jh = {
+                    let cas = cas.clone();
+
+                    tokio::spawn(async move {
+                        cas.upload_all_staged(MAX_CONCURRENT_UPLOADS, false)
+                            .await
+                            .or_else(convert_cas_error)
+                    })
+                };
+
+                // Get a list of all the merged shards in order to upload them.
+                let merged_shards = merged_shards_jh.await??;
+
+                // Now, these need to be sent to the remote.  The rest of this task can happen
+                // independently of all of the notes stuff.
+                let upload_shard_jh = {
+                    let config = self.xet_config.clone();
+                    let cas = cas.clone();
+                    tokio::spawn(async move {
+                        sync_session_shards_to_remote(&config, &cas, merged_shards).await
+                    })
+                };
+
+                // Now while the shards are being uploaded, we can process the summaries.
+                update_summaries_to_git(
+                    &self.xet_config,
+                    &self.summaries_file,
+                    GIT_NOTES_SUMMARIES_REF_NAME,
+                )
+                .await?;
+
+                // Now, make sure we finish up all the remaining tasks that are running in the background.
+                upload_shard_jh.await??;
+
+                // Write v2 ref notes.  Make sure we do this after the shards have all
+                // finished uploading, or we could get
+                // notes pushed without a corresponding object in CAS.
+                update_mdb_shards_to_git_notes(
+                    &self.xet_config,
+                    &self.merkledb_v2_session_dir,
+                    GIT_NOTES_MERKLEDB_V2_REF_NAME,
+                )?;
 
-        // the first upload staged is to ensure all xorbs are synced
-        // so shard registration (in MDBv2) in sync_dbs_to_notes will find them.
-        self.upload_all_staged().await?;
-        self.sync_dbs_to_notes().await?;
-        // the second upload staged is to ensure xorbs associated with large MDBv1
-        // diff as standalone pointer file as synced.
-        self.upload_all_staged().await?;
-        self.sync_notes_to_remote(remote)?;
+                self.sync_notes_to_remote(remote)?;
+
+                // Finally, we can move all the mdb shards from the session directory, which is used
+                // by the upload_shard task, to the cache.
+                move_session_shards_to_local_cache(
+                    &self.merkledb_v2_session_dir,
+                    &self.merkledb_v2_cache_dir,
+                )
+                .await?;
 
+                upload_all_jh.await??;
```

Comment:
Move this to before `sync_notes_to_remote` so there's no dependency issue in the remote?

Reply:
In this case, the upload all task is just uploading all of the staged CAS blocks from the files. So it is independent of any of the notes.